File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/npcs.ts
/**
 * Campaign NPCs and factions: create factions, create NPCs (optionally
 * linked to a faction), and summarize relationship counts.
 */

import { getDb } from '../db.ts';
import type { ApiResult, JsonValue } from '../types.ts';
import { isValidIntInRange } from '../validation.ts';

const FACTION_STANCES = new Set(['friendly', 'neutral', 'hostile']);

export function createFaction(campaignId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const id = body.id;
  if (typeof id !== 'string' || id.length === 0) {
    return { status: 400, body: { error: 'id must be a non-empty string' } };
  }

  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const stance = body.stance;
  if (typeof stance !== 'string' || !FACTION_STANCES.has(stance)) {
    return { status: 400, body: { error: 'stance must be one of friendly, neutral, hostile' } };
  }

  const existing = db.prepare('SELECT id FROM campaign_factions WHERE campaign_id = ? AND id = ?').get(campaignId, id);
  if (existing) {
    return { status: 409, body: { error: 'faction id already exists' } };
  }

  db.prepare('INSERT INTO campaign_factions (campaign_id, id, name, stance) VALUES (?, ?, ?, ?)').run(
    campaignId,
    id,
    name,
    stance,
  );

  return { status: 201, body: { id, name, stance } };
}

export function createNpc(campaignId: string, body: JsonValue): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const id = body.id;
  if (typeof id !== 'string' || id.length === 0) {
    return { status: 400, body: { error: 'id must be a non-empty string' } };
  }

  const name = body.name;
  if (typeof name !== 'string' || name.length === 0) {
    return { status: 400, body: { error: 'name must be a non-empty string' } };
  }

  const factionId = body.faction_id;
  if (factionId !== undefined && factionId !== null && typeof factionId !== 'string') {
    return { status: 400, body: { error: 'faction_id must be a string' } };
  }

  if (typeof factionId === 'string') {
    const faction = db
      .prepare('SELECT id FROM campaign_factions WHERE campaign_id = ? AND id = ?')
      .get(campaignId, factionId);
    if (!faction) {
      return { status: 400, body: { error: 'unknown faction id' } };
    }
  }

  const disposition = body.disposition;
  if (!isValidIntInRange(disposition, -10, 10)) {
    return { status: 400, body: { error: 'disposition must be an integer between -10 and 10' } };
  }

  const existing = db.prepare('SELECT id FROM campaign_npcs WHERE campaign_id = ? AND id = ?').get(campaignId, id);
  if (existing) {
    return { status: 409, body: { error: 'npc id already exists' } };
  }

  db.prepare(
    'INSERT INTO campaign_npcs (campaign_id, id, name, faction_id, disposition) VALUES (?, ?, ?, ?, ?)',
  ).run(campaignId, id, name, typeof factionId === 'string' ? factionId : null, disposition);

  return {
    status: 201,
    body: { id, name, faction_id: typeof factionId === 'string' ? factionId : null, disposition },
  };
}

export function relationshipSummary(campaignId: string): ApiResult {
  const db = getDb();
  const campaign = db.prepare('SELECT id FROM campaigns WHERE id = ?').get(campaignId);
  if (!campaign) {
    return { status: 404, body: { error: 'unknown campaign id' } };
  }

  const factionCount = db
    .prepare('SELECT COUNT(*) as count FROM campaign_factions WHERE campaign_id = ?')
    .get(campaignId) as unknown as { count: number };

  const npcCount = db
    .prepare('SELECT COUNT(*) as count FROM campaign_npcs WHERE campaign_id = ?')
    .get(campaignId) as unknown as { count: number };

  const friendlyCount = db
    .prepare('SELECT COUNT(*) as count FROM campaign_npcs WHERE campaign_id = ? AND disposition > 0')
    .get(campaignId) as unknown as { count: number };

  return {
    status: 200,
    body: {
      campaign_id: campaignId,
      factions: factionCount.count,
      npcs: npcCount.count,
      friendly_npcs: friendlyCount.count,
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T070752Z_claude_claude-sonnet-5_ruby-stdlib/lib/handlers/npcs.rb
# frozen_string_literal: true

require_relative '../errors'
require_relative '../database'
require_relative 'campaigns'

module Handlers
  # Campaign NPCs, factions, and a relationship summary derived from them.
  module Npcs
    STANCES = %w[friendly neutral hostile].freeze

    module_function

    def create_faction(campaign_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)

      id = body['id']
      name = body['name']
      stance = body['stance']

      raise HttpError.new(400, 'id must be a string') unless id.is_a?(String) && !id.empty?
      raise HttpError.new(400, 'name must be a string') unless name.is_a?(String) && !name.empty?
      raise HttpError.new(400, 'stance must be one of friendly, neutral, hostile') unless STANCES.include?(stance)
      raise HttpError.new(409, 'id already exists') unless Database.query("SELECT id FROM campaign_factions WHERE id = #{Database.escape(id)};").empty?

      Database.exec(<<~SQL)
        INSERT INTO campaign_factions (id, campaign_id, name, stance)
        VALUES (
          #{Database.escape(id)},
          #{Database.escape(campaign_id)},
          #{Database.escape(name)},
          #{Database.escape(stance)}
        );
      SQL

      [201, { id: id, name: name, stance: stance }]
    end

    def create_npc(campaign_id, body)
      Handlers::Campaigns.find_campaign(campaign_id)

      id = body['id']
      name = body['name']
      faction_id = body['faction_id']
      disposition = body['disposition']

      raise HttpError.new(400, 'id must be a string') unless id.is_a?(String) && !id.empty?
      raise HttpError.new(400, 'name must be a string') unless name.is_a?(String) && !name.empty?
      raise HttpError.new(400, 'disposition must be an integer') unless disposition.is_a?(Integer)
      unless faction_id.nil? || (faction_id.is_a?(String) && !faction_id.empty?)
        raise HttpError.new(400, 'faction_id must be a string')
      end
      raise HttpError.new(409, 'id already exists') unless Database.query("SELECT id FROM campaign_npcs WHERE id = #{Database.escape(id)};").empty?

      unless faction_id.nil?
        raise HttpError.new(400, 'unknown faction id') if Database.query(
          "SELECT id FROM campaign_factions WHERE campaign_id = #{Database.escape(campaign_id)} AND id = #{Database.escape(faction_id)};"
        ).empty?
      end

      Database.exec(<<~SQL)
        INSERT INTO campaign_npcs (id, campaign_id, name, faction_id, disposition)
        VALUES (
          #{Database.escape(id)},
          #{Database.escape(campaign_id)},
          #{Database.escape(name)},
          #{faction_id.nil? ? 'NULL' : Database.escape(faction_id)},
          #{Database.int(disposition)}
        );
      SQL

      [201, { id: id, name: name, faction_id: faction_id, disposition: disposition }]
    end

    def relationships(campaign_id, _body)
      Handlers::Campaigns.find_campaign(campaign_id)

      faction_count_row = Database.query("SELECT COUNT(*) AS count FROM campaign_factions WHERE campaign_id = #{Database.escape(campaign_id)};").first
      npc_count_row = Database.query("SELECT COUNT(*) AS count FROM campaign_npcs WHERE campaign_id = #{Database.escape(campaign_id)};").first
      friendly_row = Database.query(<<~SQL).first
        SELECT COUNT(*) AS count
        FROM campaign_npcs n
        JOIN campaign_factions f ON f.id = n.faction_id AND f.campaign_id = n.campaign_id
        WHERE n.campaign_id = #{Database.escape(campaign_id)} AND f.stance = 'friendly';
      SQL

      [200, {
        campaign_id: campaign_id,
        factions: faction_count_row ? faction_count_row['count'].to_i : 0,
        npcs: npc_count_row ? npc_count_row['count'].to_i : 0,
        friendly_npcs: friendly_row ? friendly_row['count'].to_i : 0
      }]
    end
  end
end
